memtable.cc

```diff
@@ -702,13 +702,29 @@ memtable::make_flat_reader(schema_ptr s,
         rd.upgrade_schema(s);
         return rd;
     } else {
-        auto res = make_flat_mutation_reader<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
-        if (fwd == streamed_mutation::forwarding::yes) {
-            return make_forwardable(std::move(res));
-        } else {
-            return res;
-        }
+        return make_flat_mutation_reader<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
+    }
+}
+
+flat_mutation_reader
+memtable::make_flat_reader(schema_ptr s, reader_permit permit, const dht::partition_range& range, const query::partition_slice& slice,
+        const io_priority_class& pc, tracing::trace_state_ptr trace_state_ptr, streamed_mutation::forwarding fwd, mutation_reader::forwarding fwd_mr) {
+    const auto reversed = slice.options.contains(query::partition_slice::option::reversed);
+    auto fwd_sm = fwd;
+    if (reversed) {
+        fwd_sm = streamed_mutation::forwarding::no;
+        s = s->make_reversed();
+    }
+
+    auto rd = do_make_flat_reader(std::move(s), permit, range, slice, pc, std::move(trace_state_ptr), fwd_sm, fwd_mr);
+
+    if (reversed) {
+        rd = make_reversing_reader(std::move(rd), permit.max_result_size());
+    }
+    if (fwd && (reversed || !query::is_single_partition(range) || fwd_mr)) {
+        rd = make_forwardable(std::move(rd));
     }
+    return rd;
 }
```

Comment:
This commit introduces an incompatibility between the memtable reader and sstable reader. See the commit message of https://github.com/scylladb/scylla/pull/9281/commits/8c036d8092c1d94c05359336751a776823f35232.

In short: the memtable reader will pass the slice down to sstable reader (when it delegates work during scanning), which is half-reversed, causing the sstable reader to perform a reverse query, but the memtable reader expects non-reversed results since reversing is done above `do_make_flat_reader`.

IIUC you can easily fix this by passing unreversed slice to `do_make_flat_reader` and storing it in reversing reader like you do with the other readers.

This problem is probably only theoretical because scanning reverse reads can only be executed from Thrift which nobody uses. (let me remind everyone that Thrift has been disabled by default in C*3)

Reply:
Oops, I think we should have a dedicated test for this scenario.

> let me remind everyone that Thrift has been disabled by default in C*3

We made a conscious decision to *not* follow suit and keep providing a haven for all the Thrift refugees.